### PR TITLE
taks_2

### PR DESCRIPTION
**Pull request by Haroon9919**
Waiting to merge task 2

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  return db.end();
+});
+
+beforeEach(() => {
+  return seed(data);
+});
+
+describe("app testing", () => {
+  describe("/api/topics", () => {
+    test("GET 200: returns with a status of 200 and an array", async () => {
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: Keep an eye on these test descriptions, you're stating here that the endpoint comes back with a 200 and an array but are only expecting a 200. Then later stating it comes back with an array. Could these be combined into one test?

File: __tests__/app.test.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      const response = await request(app).get("/api/topics");
+      const topics = response.body;
+
+      expect(topics).toBeInstanceOf(Array);
+      return topics.forEach((topic) => {
+        expect(topic).toHaveProperty("slug");
+        expect(topic).toHaveProperty("description");
+      });
+    });
+    test("GET 404: should return an error for a non-existant path", () => {
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: Is this 404 relevant to topics only? This should be in it's own describe block as this is a universal error.

File: controllers/topics.controllers.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const express = require('express');
+const app = express();
```
Comment (dappernerddesigns) [inline controllers/topics.controllers.js]: Why is this in the controller? Is your app in the controller?

File: controllers/topics.controllers.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+const topicData = require('../db/data/development-data/topics');
```
Comment (dappernerddesigns) [inline controllers/topics.controllers.js]: Why are you requiring in the data from the dev-data folder? What if you didn't have access to this information?

File: controllers/topics.controllers.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+const getTopics = app.get('/api/topics', (req, res) => {
+  const topics = topicData;
+  return res.status(200).send(topics);
+});
```
Comment (dappernerddesigns) [inline controllers/topics.controllers.js]: This is fundamentally not what to do with a controller. This function should be invoking a model function that uses psql to interact with the database.

**Review by dappernerddesigns — CHANGES_REQUESTED** [review]
There is a lot going on here Haroon that needs addressing. Fundamentally this is not how to interact with your database. You won't always have the data to be able to import and send back to a user. Your model function should be making use of psql queries to the database to get you back what you are looking for.

Can you please go back to this question and resubmit when you have correctly acquired and tested your data.

File: __tests__/app.test.js [hunk 6]
```diff
@@ -14,22 +14,24 @@ beforeEach(() => {
...
   describe("/api/topics", () => {
-    test("GET 200: returns with a status of 200 and an array", async () => {
-      await request(app).get("/api/topics").expect(200);
+    test("GET 200: should respond with an array of topics object that have slug and description properties and status 200", () => {
+      return request(app)
+        .get("/api/topics")
+        .expect(200)
+        .then((response) => {
+          const topics = response.body;
+          expect(topics).toBeInstanceOf(Array);
```
Comment (mezzDavies) [inline __tests__/app.test.js:23]: We need to check the length of the array too as this will make sure we get the expected number of topics back as well as protect against a false positive test (from the forEach not being entered) in the event of no array being returned at all. You could do this instead of line 23. 😄

**Review by mezzDavies — APPROVED** [review]
This is good to merge so long as you take care of the comment below please 👍
